metaserver/src/main/scala/scala/meta/languageserver/SymbolIndexerMap.scala

```diff
@@ -0,0 +1,87 @@
+package scala.meta.languageserver
+
+import java.util.concurrent.atomic.AtomicReference
+import java.util.function.UnaryOperator
+import scala.collection.concurrent.TrieMap
+import scala.meta.languageserver.index.Position
+import scala.meta.languageserver.index.Range
+import scala.meta.languageserver.index.Ranges
+import scala.meta.languageserver.index.SymbolIndex
+import com.typesafe.scalalogging.LazyLogging
+import org.langmeta.semanticdb.Symbol
+
+class SymbolIndexerMap(
+    symbols: TrieMap[String, AtomicReference[SymbolIndex]] = TrieMap.empty
+) extends LazyLogging { self =>
+  private def newValue(symbol: String) = {
+    new AtomicReference(SymbolIndex(symbol = symbol))
+  }
+
+  def index: Traversable[SymbolIndex] = new Traversable[SymbolIndex] {
+    override def foreach[U](f: SymbolIndex => U): Unit =
+      symbols.values.foreach(s => f(s.get))
+  }
+
+  def updated(symbol: String)(f: SymbolIndex => SymbolIndex): Unit = {
+    val value = symbols.getOrElseUpdate(symbol, newValue(symbol))
+    value.getAndUpdate(new UnaryOperator[SymbolIndex] {
```

Comment:
This too can be a SAM, I think.

I haven't tried, but this may actually be as simple as:

```scala
value.getAndUpdate(f)
```